src/impls.rs

```diff
@@ -113,200 +90,119 @@ impl_metadata_for_tuple!(A, B, C, D, E, F, G, H);
 impl_metadata_for_tuple!(A, B, C, D, E, F, G, H, I, J);
 
-impl<T> HasTypeId for Vec<T>
-where
-	T: Metadata + 'static,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("Vec", Namespace::prelude(), tuple_meta_type![T]).into()
-	}
-}
-
-impl<T> HasTypeDef for Vec<T>
+impl<T> TypeInfo for Vec<T>
 where
 	T: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("elems", MetaType::new::<[T]>())]).into()
+	fn type_info() -> Type {
+		<[T] as TypeInfo>::type_info()
 	}
 }
 
-impl<T> HasTypeId for Option<T>
+impl<T> TypeInfo for Option<T>
 where
 	T: Metadata + 'static,
 {
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("Option", Namespace::prelude(), tuple_meta_type![T]).into()
+	fn type_info() -> Type {
+		TypeVariant::new("Option", Namespace::prelude())
+			.type_params(tuple_meta_type![T])
+			.variants(
+				Variants::with_fields()
+					.variant_unit("None")
+					.variant("Some", Fields::unnamed().field_of::<T>()),
+			)
+			.into()
 	}
 }
 
-impl<T> HasTypeDef for Option<T>
-where
-	T: Metadata + 'static,
-{
-	fn type_def() -> TypeDef {
-		TypeDefEnum::new(vec![
-			EnumVariantUnit::new("None").into(),
-			EnumVariantTupleStruct::new("Some", vec![UnnamedField::of::<T>()]).into(),
-		])
-		.into()
-	}
-}
-
-impl<T, E> HasTypeId for Result<T, E>
-where
-	T: Metadata + 'static,
-	E: Metadata + 'static,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("Result", Namespace::prelude(), tuple_meta_type!(T, E)).into()
-	}
-}
-
-impl<T, E> HasTypeDef for Result<T, E>
+impl<T, E> TypeInfo for Result<T, E>
 where
 	T: Metadata + 'static,
 	E: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefEnum::new(vec![
-			EnumVariantTupleStruct::new("Ok", vec![UnnamedField::of::<T>()]).into(),
-			EnumVariantTupleStruct::new("Err", vec![UnnamedField::of::<E>()]).into(),
-		])
-		.into()
-	}
-}
-
-impl<K, V> HasTypeId for BTreeMap<K, V>
-where
-	K: Metadata + 'static,
-	V: Metadata + 'static,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("BTreeMap", Namespace::prelude(), tuple_meta_type!(K, V)).into()
+	fn type_info() -> Type {
+		TypeVariant::new("Result", Namespace::prelude())
+			.type_params(tuple_meta_type!(T, E))
+			.variants(
+				Variants::with_fields()
+					.variant("Ok", Fields::unnamed().field_of::<T>())
+					.variant("Err", Fields::unnamed().field_of::<E>()),
+			)
+			.into()
 	}
 }
 
-impl<K, V> HasTypeDef for BTreeMap<K, V>
+impl<K, V> TypeInfo for BTreeMap<K, V>
 where
 	K: Metadata + 'static,
 	V: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("elems", MetaType::new::<[(K, V)]>())]).into()
-	}
-}
-
-impl<T> HasTypeId for Box<T>
-where
-	T: HasTypeId + ?Sized,
-{
-	fn type_id() -> TypeId {
-		T::type_id()
+	fn type_info() -> Type {
+		TypeComposite::new("BTreeMap", Namespace::prelude())
+			.type_params(tuple_meta_type![(K, V)])
+			.fields(Fields::named().field_of::<[(K, V)]>("elems"))
+			.into()
```

Comment:
Can we maybe represent it like a `TupleStruct`? So that we can get rid of `elems` field?